packages/contractkit/src/providers/celo-provider.ts

```diff
@@ -1,80 +1,170 @@
-import { JSONRPCRequestPayload, Web3ProviderEngine } from '@0x/subproviders'
 import debugFactory from 'debug'
-import { JsonRPCResponse, Provider } from 'web3/providers'
-import { Callback } from 'web3/types'
-import { CeloPrivateKeysWalletProvider } from './celo-private-keys-subprovider'
-import { WrappingSubprovider } from './wrapping-subprovider'
+import { Callback, JsonRPCRequest, JsonRPCResponse, Provider } from 'web3/providers'
+import { MissingTxParamsPopulator } from '../utils/missing-tx-params-populator'
+import { stopProvider } from '../utils/provider-utils'
+import { DefaultRpcCaller, RpcCaller, rpcCallHandler } from '../utils/rpc-caller'
+import { DefaultWallet, Wallet } from '../utils/wallet'
 
-const debug = debugFactory('kit:provider:celo-provider')
+const debug = debugFactory('kit:provider:connection')
+const debugPayload = debugFactory('kit:provider:payload')
+const debugResponse = debugFactory('kit:provider:response')
 
-type OnFn = (event: string, handler: () => void) => void
 export class CeloProvider implements Provider {
-  public readonly on: null | OnFn = null
-  private readonly providerEngine: Web3ProviderEngine
-  private readonly localSigningProvider: CeloPrivateKeysWalletProvider
-  private readonly underlyingProvider: Provider
-
-  constructor(readonly existingProvider: Provider, readonly privateKey: string) {
-    debug('Setting up providers...')
-    // Create a Web3 Provider Engine
-    this.providerEngine = new Web3ProviderEngine()
-    // Compose our Providers, order matters
-    // First add provider for signing
-    this.localSigningProvider = new CeloPrivateKeysWalletProvider(privateKey)
-    this.providerEngine.addProvider(this.localSigningProvider)
-    // Use the existing provider to route all other requests
-    const wrappingSubprovider = new WrappingSubprovider(existingProvider)
-    this.providerEngine.addProvider(wrappingSubprovider)
-    this.underlyingProvider = existingProvider
-
-    // Initializer "on" conditionally.
-    if (existingProvider.hasOwnProperty('on')) {
-      this.on = (event: string, handler: () => void): void => {
-        this.providerEngine.on(event, handler)
-      }
-    }
+  private readonly wallet: Wallet
+  private readonly rpcCaller: RpcCaller
+  private readonly paramsPopulator: MissingTxParamsPopulator
+  private alreadyStopped: boolean = false
```

Comment:
Can a provider be stopped and then restarted?

Reply:
No, we are closing/destroying the underlying connection. We could eventually make a restart, if we get all the connection info before the disconnection, but I'm not sure if there is a use case where the user wants that instead of creating again the provider.
I mean, the provider has some state, so probably we will need to understand what to add or remove in a restart (and have a deeper control of the connection)
In my opinion, the cost-benefit of a restart instead of re-creating the Provider it not seems good.
One thing that we could eventually do, is to make the `wallet` used by the `CeloProvider` settable, this way you can use the same `local addresses data` for different Providers